File: latihan/2-inheritance/Mobil.cpp
/*  NIM     : 13521129
    Nama    : Chiquita Ahsanunnisa
    Tanggal : 19 Februari 2023
    Topik   : Latihan Inheritance - Mobil.cpp
*/

#include "Mobil.hpp"
// MOBIL (DERIVED CLASS)

// CTOR
Mobil::Mobil() : Kendaraan(0,0,"XXX"), supir("XXXX") {}
Mobil::Mobil(int n, int t, string m, string s) : Kendaraan(n,t,m), supir(s) {}

// CCTOR tidak butuh karena tidak ada DMA
// Mobil::Mobil(const Mobil& m) {}

// ASSIGNMENT tidak butuh karena tidak ada DMA
// Mobil& Mobil::operator=(const Mobil& m) {}

// DTOR tidak butuh karena tidak ada DMA baik di base maupun derived class
// Mobil::~Mobil() {}

// redefine virtual method
void Mobil::printInfo() const {
    Kendaraan::printInfo();
    cout << "Kategori: Mobil" << endl;
}
int Mobil::biayaSewa(int lamaSewa) const {
    return 500000 * lamaSewa;
}

// VIRTUAL COPY
Kendaraan* Mobil::clone() const {
    return new Mobil(*this);
}

File: latihan/5-java-inheritance/Mobil.java
/*  NIM     : 13521129
    Nama    : Chiquita Ahsanunnisa
    Tanggal : 31 Maret 2023
    Topik   : Latihan Java Inheritance - Mobil.java
*/

public class Mobil extends Kendaraan {
    private String supir;

    // CTOR
    public Mobil() {
        super();
        supir = "XXXX";
    }
    public Mobil(int nomor, int tahun, String merk, String supir) {
        super(nomor, tahun, merk);
        this.supir = supir;
    }
    public Mobil(Mobil other) {
        super(other.getNomor(), other.getTahun(), other.getMerk());
        this.supir = other.supir;
    }

    // OTHERS
    @Override
    public void printInfo() {
        super.printInfo();
        System.out.println("Kategori       : Mobil");
        System.out.println("Supir          : " + supir);
    }
    @Override
    public int biayaSewa(int lamaSewa) { return 500000 * lamaSewa; }
    @Override
    public Kendaraan copy() { return new Mobil(this); }
}
